### ISSUE TITLE
Instantiate a python subclass with an already-created cpp instance

### ISSUE THREAD

**Issue by dfb** [pybind/pybind11#2315]
Hi, this isn't a bug so much as looking for some direction. I'm using pybind11 to wrap classes in a 3rd party game engine. It is already designed to allow you to create your own subclasses, but one wrinkle is that you can't create objects directly (via e.g. `new MySubclass`) but instead have to call a factory function the engine provides. My goal is to create Python subclasses for this engine, but due to the way objects are instantiated, I need to figure out how to take a C++ instance of a base class and then dynamically extend it to become an instance of the Python subclass.

The engine provides an `Actor` base class, and I extend it in C++ as `MyActor`, and I set up pybind11 bindings for it, along with a trampoline class `PyMyActor` so that it can be subclassed in Python:

```c++
class Actor  // provided by the engine

class MyActor : public Actor // my subclass; I use compile-time macros to expose this class to the engine

class PyMAyctor : public MyActor // my trampoline class

PYBIND11_EMBEDDED_MODULE(engine, m) {
  py:class_<MyActor, PyMyActor>(m, "MyActor") ...

  m.def("spawn", [](py::object klass) -> py::object
  {
      MyActor *inst = engine.spawn<MyActor>();
      // TODO: "extend" this to be an instance of klass
  });
}
```

and then in Python something like:

```py
import engine
class Foo(engine.MyActor):
   pass
...
newActor = engine.spawn(Foo)
```

The engine will only ever know about my base class (and/or maybe the trampoline class, if needed) as that is all that is known about at compile time. The `spawn` function in C++ would normally just create a new instance by calling the passed in class, but that's not possible here because the engine needs to create the C++ object - so basically I need to split apart the instantiation process that pybind11 does, and only do the "stuff" after that C++ object is created (creating the corresponding Python object for the subclass).

As a hacky proof of concept, I modified `pybind11_object_new` to allow me to override the type object:

```c++
PyTypeObject *nextType = nullptr; // an unguarded global variable, wheeee
extern "C" inline PyObject *pybind11_object_new(PyTypeObject *type, PyObject *, PyObject *) {
    if (nextType != nullptr)
    {
        type = nextType;
        nextType = nullptr;
    }
    return make_new_instance(type);
}
```

and then in my `spawn` function I tried:

```c++
    py::detail::nextType = (PyTypeObject*)klass.ptr();
    PyMyActor *inst = engine.spawn<PyMyActor>();
```

but that didn't trigger the pybind11 code that instantiated the Python subclass (pybind11_object_new wasn't called AFAICT) so I'm obviously not hooking in at the right place.

So to recap, I'm looking for a way to do:

```c++
py::object klassFromPython; // provded by the caller
py::object inst = klassFromPython();
```

Except that instead of having pybind11 creating the C++ object and the Python object, I provide the C++ object and then it does the rest of the usual steps. Many thanks for any ideas/suggestions (even incomplete ones)!

**Comment by bstaletic** [comment]
You can create a python side constructor that calls that factory function. Something along these lines:

```c++
    py::class_<MyActor, PyMyActor>(m, "MyActor")
            .def("__init__", [](...args...) { return factory(...args...); });
```

**Comment by dfb** [comment]
Ahh, ok, I get it. Thanks very much!
